File: 404/sum-of-left-leaves.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *   this.val = val;
 *   this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
const sumOfLeftLeaves = root => {
  if (root === null) {
    return 0
  }
  const left = isLeaf(root.left) ? root.left.val : sumOfLeftLeaves(root.left)
  const right = sumOfLeftLeaves(root.right)
  return left + right
}

const isLeaf = node => (
  node !== null && 
  node.left === null && 
  node.right === null
)

File: 404/sum-of-left-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *   int val;
 *   TreeNode *left;
 *   TreeNode *right;
 *   TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
  private:
    bool isLeaf(TreeNode* node) {
      return (
        node != nullptr and 
        node->left == nullptr and 
        node->right == nullptr
      );
    }

  public:
    int sumOfLeftLeaves(TreeNode* root) {
      if (root == nullptr) {
        return 0;
      }
      int left = isLeaf(root->left) 
        ? root->left->val
        : sumOfLeftLeaves(root->left);
      int right = sumOfLeftLeaves(root->right);
      return left + right;
    }
};
